ldap/servers/slapd/haproxy.c

```diff
@@ -0,0 +1,376 @@
+            slapi_log_err(SLAPI_LOG_CONNS, "haproxy_parse_v1_hdr", "Missing protocol header terminator\n");
+            goto done;
+        }
+        /* Parse the protocol, family, addresses, and ports */
+        if (haproxy_parse_v1_protocol(get_next_token(&copied), "PROXY") < 0) {
+            slapi_log_err(SLAPI_LOG_CONNS, "haproxy_parse_v1_hdr", "Missing or bad protocol header\n");
+            goto done;
+        }
+        /* Parse the family */
+        if (haproxy_parse_v1_fam(get_next_token(&copied), &addr_family) < 0) {
+            slapi_log_err(SLAPI_LOG_CONNS, "haproxy_parse_v1_hdr", "Missing or bad protocol type\n");
+            goto done;
+        }
+        /* Parse the addresses */
+        if (haproxy_parse_v1_addr(get_next_token(&copied), &parsed_addr_from, addr_family) < 0) {
+            slapi_log_err(SLAPI_LOG_CONNS, "haproxy_parse_v1_hdr", "Missing or bad client address\n");
+            goto done;
+        }
+        if (haproxy_parse_v1_addr(get_next_token(&copied), &parsed_addr_dest, addr_family) < 0) {
+            slapi_log_err(SLAPI_LOG_CONNS, "haproxy_parse_v1_hdr", "Missing or bad server address\n");
+            goto done;
+        }
+        /* Parse the ports */
+        if (haproxy_parse_v1_port(get_next_token(&copied), &parsed_addr_from) < 0) {
+            slapi_log_err(SLAPI_LOG_CONNS, "haproxy_parse_v1_hdr", "Missing or bad client port\n");
+            goto done;
+        }
+        if (haproxy_parse_v1_port(get_next_token(&copied), &parsed_addr_dest) < 0) {
+            slapi_log_err(SLAPI_LOG_CONNS, "haproxy_parse_v1_hdr", "Missing or bad server port\n");
+            goto done;
+        }
+        rc = 0;
+        *proxy_connection = 1;
+        *str_len = after_header - str_saved;
+        /* Copy the parsed addresses to the output parameters */
+        memcpy(pr_netaddr_from, &parsed_addr_from, sizeof(PRNetAddr));
+        memcpy(pr_netaddr_dest, &parsed_addr_dest, sizeof(PRNetAddr));
+
+done:
+        slapi_ch_free_string(&str_saved);
+    }
+    return rc;
+}
+
+/**
+ * Function to receive and parse HAProxy headers, supporting both v1 and v2 of the protocol.
+ * 
+ * @param fd: The file descriptor of the socket from which to read.
+ * @param proxy_connection: A pointer to an integer to store the proxy connection status (0 or 1).
+ * @param pr_netaddr_from: A pointer to a PRNetAddr structure to store the source address info.
+ * @param pr_netaddr_dest: A pointer to a PRNetAddr structure to store the destination address info.
+ * 
+ * @return: Returns 0 on successful operation, -1 on error.
+ */
+int haproxy_receive(int fd, int *proxy_connection, PRNetAddr *pr_netaddr_from, PRNetAddr *pr_netaddr_dest)
+{
+    /* Buffer to store the header received from the HAProxy server */
+    char hdr[HAPROXY_HEADER_MAX_LEN + 1];
+    size_t hdr_len;
+
+    /* Attempt to receive the header from the HAProxy server */
+    if ((hdr_len = recv(fd, hdr, sizeof(hdr) - 1, MSG_PEEK)) <= 0) {
```

Comment:
@Firstyear okay, I've moved the `haproxy_recieve` call inside of `connection_reset`. The reason is that we need to set the client address correctly, and  `connection_reset` already logs the IP to the access log (so it'll be confusing to parse it later as the proxy IP would not be a correct IP printed in this case).

> First, the reading of this data can't and shouldn't be in the accept thread - the accept thread has to be as light as possible, and has to accept quickly, so this will potentially add latency and performance issues.

If I understand correctly, `connection_reset` is still in the `accept_thread`. What I can think of - we can add a config setting - something like `nsslapd-parse-haproxy: [on, off]` so the admin will use it only when they have HAProxy in their topology. This way, it will harm the performance only when they have HAProxy (but they will win in other ways). What do you think?

> Second, that the reading of this proxy header must occur in a way that a short message won't stall the accept thread, or that a timeout occurs.

`recv` won't wait on 1 byte as per specs. But it will block on 0 bytes. I've added a 'nonblocking switch' to that area. Do you think it will work for our case?

Reply:
Blocking in the accept thread isn't really something we can do, nor can we have the accept thread stuck waiting on the haproxy info. This must be in the client thread, so if that means we have to change how we log the IP address then that's a *good* thing. For example, it actually probably is good to log the connection IP as coming from the HAPROXY server, and then have a special log message that states "connection from HAPROXY from IP X:Y originates from client A:B"

Reply:
But that's exactly what I do, I unblock the socket so `recv` won't wait on anything and fails right away if there's 0 byte. 

Please, check haproxy.c and its `set_nonblocking_if_needed` and `set_blocking_if_needed` around `recv` call...

Reply:
It's still happening in the accept thread though, which is kind of my point here that it will add latency to the accept thread - there are enough issues with people trying to improve the performance of this, so I'm worried that we'll end up some issues here. 

For example, imagine if for some reason a valid haproxy header was sent but then fragments over the recv call? We would then close the client connection immediately. Or if the connection opens and 0 bytes are sent, then will we skip the haproxy header entirely and then move to the worker which will cause the initial message to be invalid BER?

I think there are a lot of ways this can go wrong in it's current design, and I'm worried about how that will impact clients in subtle ways. I think the right thing is to check for the haproxy header in the worker threads before the first message is received, and if *then* it's present and is a valid haproxy header we pivot the address of the conn. If there is no header we proceed as usual.

Reply:
Thanks! Sounds good, and I surely don't mind. :)
As I mentioned, my 389 DS connection management knowledge is sparse.
Will this be a good place to put the `haproxy_receive()` call and then continue with logging? (line 1171)
https://github.com/389ds/389-ds-base/blob/main/ldap/servers/slapd/connection.c#L1171

Is it what you meant?